dev/AppNotifications/AppNotificationTelemetry.h

```
@@ -0,0 +1,318 @@
+
+    UINT UpdateLogEventCount() noexcept
+    {
+        ULONGLONG currentTick = GetTickCount64();
+
+        auto lock{ m_lock.lock_exclusive() };
+
+        // Only fire limiting events every log period to prevent too many events from being fired
+        if ((currentTick - m_lastFiredTick) > c_logPeriod)
+        {
+            m_eventCount = 0;
+            m_lastFiredTick = currentTick;
+        }
+
+        m_eventCount++;
+
+        return m_eventCount;
+    }
+
+    inline bool IsPackagedApp()
+    {
+        static const bool isPackagedApp = AppModel::Identity::IsPackagedProcess();
+
+        return isPackagedApp;
+    }
+
+    inline const wchar_t* GetAppName()
+    {
+        static const std::wstring appName = IsPackagedApp() ? GetAppNamePackaged() : GetAppNameUnpackaged();
+
+        return appName.c_str();
+    }
+
+    std::wstring GetAppNamePackaged() noexcept
+    {
+        wchar_t appUserModelId[APPLICATION_USER_MODEL_ID_MAX_LENGTH]{};
+
+        UINT32 appUserModelIdSize = ARRAYSIZE(appUserModelId);
+        auto result = GetCurrentApplicationUserModelId(&appUserModelIdSize, appUserModelId);
+        if (result != ERROR_SUCCESS)
+        {
+            wcscpy_s(appUserModelId, L"AppUserModelId not found");
+            LOG_WIN32(result);
+        }
+
+        return appUserModelId;
+    }
+
+    PCWSTR CensorFilePath(PCWSTR path) noexcept
+    {
+        if (path)
+        {
+            path = !PathIsFileSpecW(path) ? PathFindFileNameW(path) : path;
+        }
+
+        return path;
+    }
+
+    std::wstring GetAppNameUnpackaged()
+    {
+        std::wstring appName;
+
+        wil::unique_cotaskmem_string processName;
+        auto result = wil::GetModuleFileNameExW(GetCurrentProcess(), nullptr, processName);
+        if (result == ERROR_SUCCESS)
+        {
+            appName = CensorFilePath(processName.get());
+        }
+        else
+        {
+            appName = L"ModuleFileName not found";
+            LOG_WIN32(result);
```

Should throw

---

THROW_HR_WIN32_MSG

---

See explanation above as to why lower layers need to throw. 
It won't lead to host of issues because the logging wrappers essentially swallow any errors today: they are wrapped in try catch_log

---

done